File: src/com/l2kt/gameserver/model/zone/type/CastleTeleportZone.java
package com.l2kt.gameserver.model.zone.type;

import com.l2kt.commons.random.Rnd;
import com.l2kt.gameserver.model.zone.ZoneId;
import com.l2kt.gameserver.model.zone.ZoneType;

import com.l2kt.gameserver.model.actor.Creature;
import com.l2kt.gameserver.model.actor.instance.Player;

/**
 * A zone extending {@link ZoneType} used for Mass Gatekeepers to teleport players on a specific location.<br>
 * <br>
 * Summoning is forbidden. It holds a location under an int array, and castleId.
 */
public class CastleTeleportZone extends ZoneType
{
	private final int[] _spawnLoc;
	private int _castleId;
	
	public CastleTeleportZone(int id)
	{
		super(id);
		
		_spawnLoc = new int[5];
	}
	
	@Override
	public void setParameter(String name, String value)
	{
		if (name.equals("castleId"))
			_castleId = Integer.parseInt(value);
		else if (name.equals("spawnMinX"))
			_spawnLoc[0] = Integer.parseInt(value);
		else if (name.equals("spawnMaxX"))
			_spawnLoc[1] = Integer.parseInt(value);
		else if (name.equals("spawnMinY"))
			_spawnLoc[2] = Integer.parseInt(value);
		else if (name.equals("spawnMaxY"))
			_spawnLoc[3] = Integer.parseInt(value);
		else if (name.equals("spawnZ"))
			_spawnLoc[4] = Integer.parseInt(value);
		else
			super.setParameter(name, value);
	}
	
	@Override
	protected void onEnter(Creature character)
	{
		character.setInsideZone(ZoneId.NO_SUMMON_FRIEND, true);
	}
	
	@Override
	protected void onExit(Creature character)
	{
		character.setInsideZone(ZoneId.NO_SUMMON_FRIEND, false);
	}
	
	public void oustAllPlayers()
	{
		if (_characters.isEmpty())
			return;
		
		for (Player player : getKnownTypeInside(Player.class))
			player.teleToLocation(Rnd.INSTANCE.get(_spawnLoc[0], _spawnLoc[1]), Rnd.INSTANCE.get(_spawnLoc[2], _spawnLoc[3]), _spawnLoc[4], 0);
	}
	
	public int getCastleId()
	{
		return _castleId;
	}
}

File: src/com/l2kt/gameserver/model/zone/type/CastleTeleportZone.kt
package com.l2kt.gameserver.model.zone.type

import com.l2kt.commons.random.Rnd
import com.l2kt.gameserver.model.actor.Creature
import com.l2kt.gameserver.model.actor.instance.Player
import com.l2kt.gameserver.model.zone.ZoneId
import com.l2kt.gameserver.model.zone.ZoneType

/**
 * A zone extending [ZoneType] used for Mass Gatekeepers to teleport players on a specific location.<br></br>
 * <br></br>
 * Summoning is forbidden. It holds a location under an int array, and castleId.
 */
class CastleTeleportZone(id: Int) : ZoneType(id) {
    private val _spawnLoc: IntArray = IntArray(5)
    var castleId: Int = 0
        private set

    override fun setParameter(name: String, value: String) {
        when (name) {
            "castleId" -> castleId = value.toInt()
            "spawnMinX" -> _spawnLoc[0] = value.toInt()
            "spawnMaxX" -> _spawnLoc[1] = value.toInt()
            "spawnMinY" -> _spawnLoc[2] = value.toInt()
            "spawnMaxY" -> _spawnLoc[3] = value.toInt()
            "spawnZ" -> _spawnLoc[4] = value.toInt()
            else -> super.setParameter(name, value)
        }
    }

    override fun onEnter(character: Creature) {
        character.setInsideZone(ZoneId.NO_SUMMON_FRIEND, true)
    }

    override fun onExit(character: Creature) {
        character.setInsideZone(ZoneId.NO_SUMMON_FRIEND, false)
    }

    fun oustAllPlayers() {
        if (_characters.isEmpty())
            return

        for (player in getKnownTypeInside(Player::class.java))
            player.teleToLocation(Rnd[_spawnLoc[0], _spawnLoc[1]], Rnd[_spawnLoc[2], _spawnLoc[3]], _spawnLoc[4], 0)
    }
}